server/schedulers/scheduler_controller.go

```diff
@@ -0,0 +1,252 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedulers
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/pd/pkg/logutil"
+	"github.com/pingcap/pd/server/core"
+	"github.com/pingcap/pd/server/schedule"
+	"github.com/pingcap/pd/server/schedule/operator"
+	"github.com/pingcap/pd/server/schedule/opt"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+const maxScheduleRetries = 10
+
+var (
+	// ErrSchedulerExisted is error info for scheduler is existed.
+	ErrSchedulerExisted = errors.New("scheduler existed")
+	// ErrSchedulerNotFound is error info for scheduler is not found.
+	ErrSchedulerNotFound = errors.New("scheduler not found")
+)
+
+// Options for SchedulerController.
+type Options interface {
+	AddSchedulerCfg(tp string, args []string)
+	RemoveSchedulerCfg(ctx context.Context, name string) error
+
+	Persist(storage *core.Storage) error
+}
+
+// SchedulerController is used to manage a scheduler to schedule.
+type SchedulerController struct {
+	sync.RWMutex
+
+	wg           sync.WaitGroup
+	schedulers   map[string]*Scheduler
+	opt          Options
+	cluster      opt.Cluster
+	storage      *core.Storage
+	opController *schedule.OperatorController
+	ctx          context.Context
+	cancel       context.CancelFunc
+}
+
+// NewSchedulerController creates a new scheduleController.
+func NewSchedulerController(ctx context.Context, cluster opt.Cluster, opController *schedule.OperatorController, opt Options, s *core.Storage) *SchedulerController {
+	ctx, cancel := context.WithCancel(ctx)
+	return &SchedulerController{
+		schedulers:   make(map[string]*Scheduler),
```

Comment:
if we use name as scheduler key, furthermore can we change scheduler to singleton?